opendevin/server/session/agent.py

```diff
@@ -0,0 +1,110 @@
+            raise Exception(
+                'Session already started. You need to close this session and start a new one.'
+            )
+        await self._create_runtime()
+        await self._create_controller(start_event)
+
+    async def close(self):
+        if self.controller is not None:
+            end_state = self.controller.get_state()
+            end_state.save_to_session(self.sid)
+            await self.controller.close()
+        if self.runtime is not None:
+            self.runtime.close()
+
+    async def _create_runtime(self):
+        if self.runtime is not None:
+            raise Exception('Runtime already created')
+        if config.runtime == 'server':
+            logger.info('Using server runtime')
+            self.runtime = ServerRuntime(self.event_stream, self.sid)
+        elif config.runtime == 'e2b':
+            logger.info('Using E2B runtime')
+            self.runtime = E2BRuntime(self.event_stream, self.sid)
+
+    async def _create_controller(self, start_event: dict):
+        """Creates an AgentController instance.
+
+        Args:
+            start_event: The start event data (optional).
+        """
+        if self.controller is not None:
+            raise Exception('Controller already created')
+        if self.runtime is None:
+            raise Exception('Runtime must be initialized before the agent controller')
+        args = {
+            key: value
+            for key, value in start_event.get('args', {}).items()
+            if value != ''
+        }  # remove empty values, prevent FE from sending empty strings
+        agent_cls = args.get(ConfigType.AGENT, config.agent.name)
+        model = args.get(ConfigType.LLM_MODEL, config.llm.model)
+        api_key = args.get(ConfigType.LLM_API_KEY, config.llm.api_key)
+        api_base = config.llm.base_url
+        max_iterations = args.get(ConfigType.MAX_ITERATIONS, config.max_iterations)
+        max_chars = args.get(ConfigType.MAX_CHARS, config.llm.max_chars)
+
+        logger.info(f'Creating agent {agent_cls} using LLM {model}')
+        llm = LLM(model=model, api_key=api_key, base_url=api_base)
+        agent = Agent.get_cls(agent_cls)(llm)
+        if isinstance(agent, CodeActAgent):
+            if not self.runtime or not isinstance(self.runtime.sandbox, DockerSSHBox):
+                logger.warning(
+                    'CodeActAgent requires DockerSSHBox as sandbox! Using other sandbox that are not stateful (LocalBox, DockerExecBox) will not work properly.'
+                )
+        self.runtime.init_sandbox_plugins(agent.sandbox_plugins)
+
+        self.controller = AgentController(
+            sid=self.sid,
+            event_stream=self.event_stream,
+            agent=agent,
+            max_iterations=int(max_iterations),
+            max_chars=int(max_chars),
+        )
+        try:
+            agent_state = State.restore_from_session(self.sid)
+            self.controller.set_state(agent_state)
```

Comment:
Can we remove `set_state` method, and initialize the state in the constructor (like in https://github.com/OpenDevin/OpenDevin/pull/1910)?

A setter method for `state` is concerning because people might abuse it.